### PR TITLE
Prepend first statement of using block with a newline, if opening bracket is on same line as using statement

### LINKED ISSUE #52970: IDE0063 does not keep line breaks and indents correctly
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/t/IDE0063-does-not-keep-line-breaks-and-in/1404759)._

---
[severity:I'm unable to use this version]
When converting a block using statement to a simple using statement, Visual Studio delete next line break.
[using.png] (https://aka.ms/dc/file?name=B457d03b0d3d441e99acfe730f34e4737637546476486574335_using.png&tid=457d03b0d3d441e99acfe730f34e4737637546476486574335)

Also, the indentation of the code in the block may be incorrect after conversion. I'm not sure what the conditions are for this to happen.

---
#### Feedback Bot on 4/21/2021, 06:59 PM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

#### Feedback Bot on 4/27/2021, 10:44 AM: 

<p>Thank you for sharing your feedback! Our teams prioritize action on product issues with broad customer impact. See details at: <a target='_blank' href="https://docs.microsoft.com/en-us/visualstudio/ide/report-a-problem?view=vs-2019#faq">https://docs.microsoft.com/en-us/visualstudio/ide/report-a-problem?view=vs-2019#faq</a>. In case you need answers to common questions or need assisted support, be sure to use <a target='_blank' href="https://visualstudio.microsoft.com/vs/support/">https://visualstudio.microsoft.com/vs/support/</a>. We’ll keep you posted on any updates to this feedback.</p>

---
### Original Solutions
(no solutions)

### PR DESCRIPTION
**Pull request by zanseb**
Prepend first statement of using block with a newline, if opening bracket is on same line as using statement

Fixes #52970

### CODE REVIEW

File: src/Analyzers/CSharp/Tests/UseSimpleUsingStatement/UseSimpleUsingStatementTests.cs [hunk 1]
```diff
@@ -1496,6 +1496,93 @@ void M()
...
+        using var a = b;
+        Console.WriteLine(a);
+    }
+}",
+parseOptions: CSharp8ParseOptions);
+        }
+
+        [WorkItem(52970, "https://github.com/dotnet/roslyn/issues/52970")]
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsUseSimpleUsingStatement)]
+        public async Task TestWithBlockBodyWithOpeningBracketAndCommentOnSameLine()
```
Comment (sharwell) [inline src/Analyzers/CSharp/Tests/UseSimpleUsingStatement/UseSimpleUsingStatementTests.cs:1562]: 💡 Maybe add a test for this?

```
using (var a = b) {
}
```

Comment (sharwell) [inline src/Analyzers/CSharp/Tests/UseSimpleUsingStatement/UseSimpleUsingStatementTests.cs:1562]: 💡 Maybe add a test for this?

```
using (var a = b) {
    // intentionally empty
}
```

File: src/Analyzers/CSharp/CodeFixes/UseSimpleUsingStatement/UseSimpleUsingStatementCodeFixProvider.cs [hunk 2]
```diff
@@ -119,8 +119,21 @@ private static SyntaxTriviaList Expand(List<StatementSyntax> result, UsingStatem
             {
                 case BlockSyntax blockSyntax:
                     var statements = blockSyntax.Statements;
+                    if (statements.Count() == 0)
```
Comment (sharwell) [inline src/Analyzers/CSharp/CodeFixes/UseSimpleUsingStatement/UseSimpleUsingStatementCodeFixProvider.cs:122]:
```suggestion
                    if (!statements.Any())
```

File: src/Analyzers/CSharp/CodeFixes/UseSimpleUsingStatement/UseSimpleUsingStatementCodeFixProvider.cs [hunk 3]
```diff
@@ -119,8 +119,21 @@ private static SyntaxTriviaList Expand(List<StatementSyntax> result, UsingStatem
...
+                        return blockSyntax.CloseBraceToken.LeadingTrivia;
+                    }
 
                     var openBraceTrailingTrivia = blockSyntax.OpenBraceToken.TrailingTrivia;
+                    var usingHasNotEndOfLineTrivia = usingStatement.CloseParenToken.TrailingTrivia
+                        .Count(t => t.IsKind(SyntaxKind.EndOfLineTrivia)) == 0;
+                    if (usingHasNotEndOfLineTrivia)
+                    {
+                        var newFirstStatement = statements.First()
+                            .WithPrependedLeadingTrivia(CSharpSyntaxFacts.Instance.ElasticCarriageReturnLineFeed);
```
Comment (sharwell) [inline src/Analyzers/CSharp/CodeFixes/UseSimpleUsingStatement/UseSimpleUsingStatementCodeFixProvider.cs:133]:
```suggestion
                            .WithPrependedLeadingTrivia(SyntaxFactory.ElasticCarriageReturnLineFeed);
```

File: src/Analyzers/CSharp/CodeFixes/UseSimpleUsingStatement/UseSimpleUsingStatementCodeFixProvider.cs [hunk 4]
```diff
@@ -119,8 +119,21 @@ private static SyntaxTriviaList Expand(List<StatementSyntax> result, UsingStatem
...
                     var statements = blockSyntax.Statements;
+                    if (statements.Count() == 0)
+                    {
+                        return blockSyntax.CloseBraceToken.LeadingTrivia;
+                    }
 
                     var openBraceTrailingTrivia = blockSyntax.OpenBraceToken.TrailingTrivia;
+                    var usingHasNotEndOfLineTrivia = usingStatement.CloseParenToken.TrailingTrivia
+                        .Count(t => t.IsKind(SyntaxKind.EndOfLineTrivia)) == 0;
+                    if (usingHasNotEndOfLineTrivia)
```
Comment (sharwell) [inline src/Analyzers/CSharp/CodeFixes/UseSimpleUsingStatement/UseSimpleUsingStatementCodeFixProvider.cs:130]:
```suggestion
                    var usingHasNoEndOfLineTrivia = !usingStatement.CloseParenToken.TrailingTrivia
                        .Any(SyntaxKind.EndOfLineTrivia);
                    if (usingHasNoEndOfLineTrivia)
```

**Review by sharwell — CHANGES_REQUESTED** [review]
Overall seems like a fine approach. Implementation has a small amount of specific suggested cleanup. Current tests are good, but I'd like to see two more added.

**Comment by zanseb** [comment]
Thank you for your feedback @sharwell, I applied your suggestions.

File: src/Analyzers/CSharp/CodeFixes/UseSimpleUsingStatement/UseSimpleUsingStatementCodeFixProvider.cs [hunk 5]
```diff
@@ -119,8 +119,21 @@ private static SyntaxTriviaList Expand(List<StatementSyntax> result, UsingStatem
...
                     var statements = blockSyntax.Statements;
+                    if (!statements.Any())
+                    {
+                        return blockSyntax.CloseBraceToken.LeadingTrivia;
+                    }
 
                     var openBraceTrailingTrivia = blockSyntax.OpenBraceToken.TrailingTrivia;
+                    var usingHasNotEndOfLineTrivia = !usingStatement.CloseParenToken.TrailingTrivia
+                        .Any(SyntaxKind.EndOfLineTrivia); ;
+                    if (usingHasNotEndOfLineTrivia)
```
Comment (sharwell) [inline src/Analyzers/CSharp/CodeFixes/UseSimpleUsingStatement/UseSimpleUsingStatementCodeFixProvider.cs:130]:
```suggestion
                    var usingHasNoEndOfLineTrivia = !usingStatement.CloseParenToken.TrailingTrivia
                        .Any(SyntaxKind.EndOfLineTrivia);
                    if (usingHasNoEndOfLineTrivia)
```

**Review by sharwell — APPROVED** [review]

File: src/Analyzers/CSharp/CodeFixes/UseSimpleUsingStatement/UseSimpleUsingStatementCodeFixProvider.cs [hunk 6]
```diff
@@ -119,8 +119,21 @@ private static SyntaxTriviaList Expand(List<StatementSyntax> result, UsingStatem
...
                     var statements = blockSyntax.Statements;
+                    if (!statements.Any())
+                    {
+                        return blockSyntax.CloseBraceToken.LeadingTrivia;
+                    }
 
                     var openBraceTrailingTrivia = blockSyntax.OpenBraceToken.TrailingTrivia;
+                    var usingHasNoEndOfLineTrivia = !usingStatement.CloseParenToken.TrailingTrivia
+                        .Any(SyntaxKind.EndOfLineTrivia);
+                    if (usingHasNoEndOfLineTrivia)
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/CodeFixes/UseSimpleUsingStatement/UseSimpleUsingStatementCodeFixProvider.cs:130]:
```suggestion
                    var hasEndOfLineTrivia = usingStatement.CloseParenToken.TrailingTrivia.Any(SyntaxKind.EndOfLineTrivia);
                    if (!hasEndOfLineTrivia)
```

**Review by CyrusNajmabadi — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
@zanseb i made a small suggestion.  LMK if you want to take it.  If not, i'm fine merging this in as-is.

**Comment by zanseb** [comment]
@CyrusNajmabadi I was thinking about this change too. Applied your suggestion, thanks!

**Comment by CyrusNajmabadi** [comment]
awesome.  thanks for the contribution!

**Pull request merged**